File: src/solver/subroutines/solveFinalCorners.ts
import { positionMap } from '../../utils'
import RubiksCubeSolver from '../RubiksCubeSolver'

export default async function solveFinalCorners(solver: RubiksCubeSolver) {
  const frontRight = solver.cubeMap.get(positionMap[9])
  const backLeft = solver.cubeMap.get(positionMap[1])
  const backRight = solver.cubeMap.get(positionMap[3])
  const frontLeft = solver.cubeMap.get(positionMap[7])

  if (frontRight?.orientation.front === frontLeft?.orientation.front) {
    await solver.do('XCW', 'XCW')
    await runAlgorithm(solver)
    return
  }

  if (frontLeft?.orientation.left === backLeft?.orientation.left) {
    await solver.do('XCW')
    await runAlgorithm(solver)
    return
  }

  if (backLeft?.orientation.back === backRight?.orientation.back) {
    await runAlgorithm(solver)
    return
  }

  if (backRight?.orientation.right === frontRight?.orientation.right) {
    await solver.do('XCCW')
    await runAlgorithm(solver)
    return
  }

  while (
    solver.cubeMap.get(positionMap[1])?.orientation.left !==
      solver.cubeMap.get(positionMap[13])?.orientation.left &&
    solver.cubeMap.get(positionMap[1])?.orientation.back !==
      solver.cubeMap.get(positionMap[11])?.orientation.back
  ) {
    await solver.do('XCW')
  }

  await runAlgorithm(solver)
  return
}

async function runAlgorithm(solver: RubiksCubeSolver) {
  await solver.do(
    'rotateRightCCW',
    'rotateFrontCW',
    'rotateRightCCW',
    'rotateBackCCW',
    'rotateBackCCW',
    'rotateRightCW',
    'rotateFrontCCW',
    'rotateRightCCW',
    'rotateBackCCW',
    'rotateBackCCW',
    'rotateRightCW',
    'rotateRightCW'
  )
}

File: android/core/src/main/kotlin/com/mikeb/simplepuzzlecube/solver/subroutines/solveFinalCorners.kt
// Port of ../src/solver/subroutines/solveFinalCorners.ts.
package com.mikeb.simplepuzzlecube.solver.subroutines

import com.mikeb.simplepuzzlecube.cubesShareFace
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateBackCCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateFrontCCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateFrontCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateRightCCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateRightCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateTopCW
import com.mikeb.simplepuzzlecube.engine.OrientationKey
import com.mikeb.simplepuzzlecube.engine.Rotation.XCCW
import com.mikeb.simplepuzzlecube.engine.Rotation.XCW
import com.mikeb.simplepuzzlecube.solver.RubiksCubeSolver

suspend fun solveFinalCorners(solver: RubiksCubeSolver) {
    val frontRight = solver.fetchPosition(9)
    val backLeft = solver.fetchPosition(1)
    val backRight = solver.fetchPosition(3)
    val frontLeft = solver.fetchPosition(7)

    if (cubesShareFace(OrientationKey.front, frontRight, frontLeft)) {
        solver.`do`(XCW, XCW)
        runAlgorithm(solver)
        return
    }

    if (cubesShareFace(OrientationKey.left, frontLeft, backLeft)) {
        solver.`do`(XCW)
        runAlgorithm(solver)
        return
    }

    if (cubesShareFace(OrientationKey.back, backLeft, backRight)) {
        runAlgorithm(solver)
        return
    }

    if (cubesShareFace(OrientationKey.right, backRight, frontRight)) {
        solver.`do`(XCCW)
        runAlgorithm(solver)
        return
    }

    while (
        !cubesShareFace(OrientationKey.left, solver.fetchPosition(1), solver.fetchPosition(13)) &&
        !cubesShareFace(OrientationKey.back, solver.fetchPosition(1), solver.fetchPosition(11))
    ) {
        solver.`do`(rotateTopCW, XCCW)
    }

    runAlgorithm(solver)
    return
}

private suspend fun runAlgorithm(solver: RubiksCubeSolver) {
    solver.`do`(
        rotateRightCCW,
        rotateFrontCW,
        rotateRightCCW,
        rotateBackCCW,
        rotateBackCCW,
        rotateRightCW,
        rotateFrontCCW,
        rotateRightCCW,
        rotateBackCCW,
        rotateBackCCW,
        rotateRightCW,
        rotateRightCW
    )
}
